sigstore-java/src/main/java/dev/sigstore/fulcio/client/FulcioCertificateMatcher.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2024 The Sigstore Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.sigstore.fulcio.client;
+
+import dev.sigstore.VerificationOptions.CertificateMatcher;
+import dev.sigstore.VerificationOptions.UncheckedCertificateException;
+import dev.sigstore.strings.StringMatcher;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateParsingException;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.logging.Logger;
+import org.bouncycastle.asn1.ASN1Primitive;
+import org.bouncycastle.asn1.ASN1Sequence;
+import org.bouncycastle.asn1.ASN1String;
+import org.bouncycastle.asn1.DEROctetString;
+import org.bouncycastle.asn1.x509.GeneralName;
+import org.bouncycastle.util.encoders.Hex;
+import org.immutables.value.Value.Immutable;
+
+@Immutable
+public abstract class FulcioCertificateMatcher implements CertificateMatcher {
+
+  public abstract StringMatcher getIssuer();
+
+  public abstract StringMatcher getSubjectAlternativeName();
+
+  /** For OIDs with raw string entries */
+  public abstract Map<String, StringMatcher> getOidRawStrings();
+
+  /** For OIDs with DER encoded UTF8 string entries */
+  public abstract Map<String, StringMatcher> getOidDerUtf8Strings();
+
+  /**
+   * For comparing raw bytes of the full ASN.1 object extension value as defined by EXTENSION in <a
+   * href="https://www.rfc-editor.org/rfc/rfc5280#section-4.1">rfc5280</a>
+   */
+  public abstract Map<String, byte[]> getOidBytes();
+
+  private static final Logger log = Logger.getLogger(FulcioCertificateMatcher.class.getName());
```

Comment:
Most of what's below this line was pulled directly from the old FulcioCertificateVerifier class, now that we've kinda standardized on matcher